### PR TITLE
 Issue #262 - Replaced JSON patch with strict update payload schema + other fixes

### PR DESCRIPTION
**Pull request by thehenrytsai**
- Issue #262 - Replaced JSON patch with strict update payload schema.
- Issue #262 - Documented update patch schema.
- Issue #172 - Removed concept of 'operation number' in update operation payload.
- Issue #120 - Added schema validation for update operation payload.
- Core service - Throwing more specific errors when operation fails validation.
- Core service - Added error code for malformed bad requests.
 - Tests - Fixed request generator used for load testing.

### CODE REVIEW

File: lib/common/ErrorCode.ts [hunk 1]
```diff
@@ -18,7 +18,29 @@ enum ErrorCode {
...
+  OperationHeaderMissingKid = 'operation_header_missing_kid',
+  OperationHeaderMissingOrIncorrectAlg = 'operation_header_missing_or_incorrect_alg',
+  OperationHeaderMissingOrIncorrectOperation = 'operation_header_missing_or_incorrect_operation',
+  OperationMissingOrIncorrectPayload = 'operation_missing_or_incorrect_payload',
+  OperationMissingOrIncorrectSignature = 'operation_missing_or_incorrect_signature',
+  OperationUpdatePayloadDidUniqueSuffixInvalid = 'operation_update_payload_did_unique_suffix_invalid',
+  OperationUpdatePayloadPreviousOperationHashInvalid = 'operation_update_payload_previous_operation_hash_invalid',
+  OperationUpdatePayloadMissingOrInvalidDidUniqueSuffixType = 'operation_update_payload_missing_or_invalid_did_unique_suffix_type',
+  OperationUpdatePayloadMissingOrInvalidPreviousOperationHashType = 'operation_update_payload_missing_or_invalid_previous_operation_hash_type',
+  OperationUpdatePayloadMissingOrUnknownProperty = 'operation_update_missing_or_unknown_property',
```
Comment (gjgd) [inline lib/common/ErrorCode.ts]: Shouldn't this be: `operation_update_payload_missing_or_unknown_property`

File: lib/common/ErrorCode.ts [hunk 2]
```diff
@@ -18,7 +18,29 @@ enum ErrorCode {
...
+  OperationUpdatePatchesNotArray = 'operation_update_patches_not_array',
+  OperationUpdatePatchMissingOrUnknownAction = 'operation_update_patch_missing_or_unknown_action',
+  OperationUpdatePatchMissingOrUnknownProperty = 'operation_update_patch_missing_or_unknown_property',
+  OperationUpdatePatchPublicKeyHexMissingOrIncorrect = 'operation_update_patch_public_key_hex_missing_or_incorrect',
+  OperationUpdatePatchPublicKeyIdNotString = 'operation_update_path_public_key_id_not_string',
```
Comment (gjgd) [inline lib/common/ErrorCode.ts]: operation_update_patch_public_key_id_not_string *

File: lib/core/Did.ts [hunk 3]
```diff
@@ -25,11 +25,22 @@ export default class Did {
...
+    if (!did.startsWith('did:')) {
+      return false;
+    }
+
+    const uniqueSuffixWithMethodName = did.substring(4); // e.g. 'sidetree:abc'
+    const indexOfColonChar = uniqueSuffixWithMethodName.indexOf(':');
+
+    // ':' must exists and not be the first or last character.
+    if (indexOfColonChar <= 0 ||
+        indexOfColonChar === (uniqueSuffixWithMethodName.length - 1)) {
```
Comment (gjgd) [inline lib/core/Did.ts]: According to the spec, `did:example:` is a valid did, but it would fail this test because ':' is tha last character

https://w3c-ccg.github.io/did-spec/#generic-did-syntax

File: tests/core/Did.spec.ts [hunk 4]
```diff
@@ -0,0 +1,8 @@
+import Did from '../../lib/core/Did';
+
+describe('DID', async () => {
+  it('isDid() should return false if DID method is not in the DID.', async () => {
+    const isDid = Did.isDid('did:abcdefg');
+    expect(isDid).toBeFalsy();
+  });
```
Comment (gjgd) [inline tests/core/Did.spec.ts]: could add a test for a string that fails the first condition of isDid() like 'abcdefg'

File: tests/core/Did.spec.ts [hunk 5]
```diff
@@ -0,0 +1,8 @@
...
+});
```
Comment (gjgd) [inline tests/core/Did.spec.ts]: coud add a positive test showing that 'did:abc:def' is a Did

File: tests/core/Operation.spec.ts [hunk 6]
```diff
@@ -34,4 +36,220 @@ describe('Operation', async () => {
...
+
+    it('should throw error if `id` of a public key in an add-public-keys patch is not a string.', async () => {
+      const updatePayload = generateUpdatePayloadForPublicKeys();
+      (updatePayload.patches[0].publicKeys![0] as any).id = { invalidType: true };
+
+      const expectedError = new SidetreeError(ErrorCode.OperationUpdatePatchPublicKeyIdNotString);
+      expect(() => { Operation.validateUpdatePayload(updatePayload); }).toThrow(expectedError);
+    });
+
+    it('should throw error if the a secp256k1 public key in an add-public-keys patch is not in `publicKeyHex` format.', async () => {
```
Comment (OR13) [inline tests/core/Operation.spec.ts]: I know we are using compressed public key format, we should maybe add a test here to ensure that the none compressed public key format won't be added, because that could cause some strange errors later...

In particular, ethereum users are likely to try and register uncompressed keys since, they are used to calculate addresesses normally...

https://bitcoin.stackexchange.com/questions/44260/can-ethereum-support-compressed-secp256k1-public-keys-to-compute-keccak-addresse

File: tests/generators/VegetaLoadGenerator.ts [hunk 7]
```diff
@@ -42,12 +43,11 @@ export default class VegetaLoadGenerator {
       fs.writeFileSync(absoluteFolderPath + `/requests/create${i}.json`, createOperationBuffer);
 
       // Compute the DID unique suffix from the generated Create payload.
-      const didUniqueSuffix = Did.getUniqueSuffixFromEncodeDidDocument(createPayload, ProtocolParameters.get(1000000).hashAlgorithmInMultihashCode);
+      const didUniqueSuffix = Did.getUniqueSuffixFromEncodeDidDocument(createPayload, ProtocolParameters.get(blockchainTime).hashAlgorithmInMultihashCode);
```
Comment (OR13) [inline tests/generators/VegetaLoadGenerator.ts]: I don't see tests in this PR for this, but I see there is an open issue for handling it later: https://github.com/decentralized-identity/sidetree/issues/265

File: lib/core/Did.ts (hunk 3 above)
Comment (thehenrytsai) [reply lib/core/Did.ts]: Won't fix unless there is a compelling reason. For the context of Sidetree, any DID without a unique DID suffix is invalid.

File: tests/core/Did.spec.ts (hunk 4 above)
Comment (thehenrytsai) [reply tests/core/Did.spec.ts]: That condition is tested already by test `should throw error if any of the service endpoints in the add-service-endpoints patch is not a valid DID.` DID.ts has 100% code coverage.

File: tests/core/Did.spec.ts (hunk 5 above)
Comment (thehenrytsai) [reply tests/core/Did.spec.ts]: DID.ts has 100% code coverage. The positive case has been hit 128 times by the tests.

File: tests/generators/VegetaLoadGenerator.ts (hunk 7 above)
Comment (thehenrytsai) [reply tests/generators/VegetaLoadGenerator.ts]: #265 is for something else, this change is more generic, it is simply removing a hardcoded value.

I think what you are looking for is a way to run load test using this tool, which is tracked by #119, though I am still guity for not doing it because it is not consider "mainnet" blocking.

File: tests/core/Operation.spec.ts (hunk 6 above)
Comment (thehenrytsai) [reply tests/core/Operation.spec.ts]: Will add a length check of 66 chars.

File: tests/generators/VegetaLoadGenerator.ts (hunk 7 above)
Comment (OR13) [reply tests/generators/VegetaLoadGenerator.ts]: yes, basically want to understand how to implement protocol upgrades, and have unit / integration tests around that.
